dot/core/service.go

```diff
@@ -204,12 +211,17 @@ func NewService(cfg *Config) (*Service, error) {
 			transactionQueue: cfg.TransactionQueue,
 			isBabeAuthority:  false,
 			lock:             chanLock,
-			closed:           false,
 			syncLock:         syncerLock,
 			blockNumOut:      cfg.SyncChan,
 			respOut:          respChan,
 		}
 
+		// thread safe way to change closed status
+		canLock := atomic.CompareAndSwapUint32(&srv.closed, 0, 1)
```

Comment:
`srv.closed` doesn't need to be updated in `NewService`, it should be updated in `Start` and `Stop`

also, `srv.closed` should be set to 0, not 1, if starting the service

Reply:
Yes, I agree that the srv.closed should be inside the Start, but this is a design change. It was like this before I started. Besides, the following tests are going to require changes since they are then failing:

```
--- FAIL: TestService_ProcessBlockAnnounceMessage (0.90s)
    messages_test.go:345: 
        	Error Trace:	messages_test.go:345
        	Error:      	Expected nil, but got: &errors.errorString{s:"[core] Error when trying to change Service status from stopped to started"}
        	Test:       	TestService_ProcessBlockAnnounceMessage

--- FAIL: TestStartService (0.94s)
    service_test.go:123: 
        	Error Trace:	service_test.go:123
        	Error:      	Expected nil, but got: &errors.errorString{s:"[core] Error when trying to change Service status from stopped to started"}
        	Test:       	TestStartService

--- FAIL: TestAnnounceBlock (0.44s)
    service_test.go:151: 
        	Error Trace:	service_test.go:151
        	Error:      	Expected nil, but got: &errors.errorString{s:"[core] Error when trying to change Service status from stopped to started"}
        	Test:       	TestAnnounceBlock

--- FAIL: TestCheckForRuntimeChanges (0.93s)
    service_test.go:210: 
        	Error Trace:	service_test.go:210
        	Error:      	Expected nil, but got: &errors.errorString{s:"service has been stopped"}
        	Test:       	TestCheckForRuntimeChanges

--- FAIL: TestAuthorModule_SubmitExtrinsic (1.73s)
    author_test.go:75: 
        	Error Trace:	author_test.go:75
        	Error:      	Expected nil, but got: &errors.errorString{s:"service has been stopped"}
        	Test:       	TestAuthorModule_SubmitExtrinsic
```

This will cause the scope will creep even more. Are you sure you want this as part of this defect ?

Reply:
okay, it's fine as is then